File: 1024-video-stitching/1024-video-stitching.cpp
#define inf 0x3f3f3f3f
class Solution {
public:
    vector<vector<int>> clips;
    int memo[101][101];
    int dp(int time, int n){
        
        if(n==-1){
            if(time==0)return 0;
            else return inf;
        }
        
        if(memo[time][n]!=-1)return memo[time][n];
        int ans=inf;
        vector<int> clip=clips[n];
        if(clip[1]>=time){ // i can pick it //
            ans=min(ans, 1+dp(clip[0], n-1));
        }
        
        ans=min(ans, dp(time, n-1));
        
        return memo[time][n]=ans;
    }
    int videoStitching(vector<vector<int>>& clips, int time) {
        
        
        memset(memo, -1, sizeof memo);
        sort(clips.begin(), clips.end(), [](vector<int> &left, vector<int>& right){
            
            if(left[0]==right[0])return left[1]<right[1];
            
            return left[0]<right[0];
        });
        
        Solution::clips=clips;
        int n=clips.size();
        int res=dp(time, n-1);
        
        if(res==inf)return -1;
        else return res;
    }
};

File: 1024-video-stitching/1024-video-stitching.py
class Solution:
    
    def videoStitching(self, clips: List[List[int]], time: int) -> int:
        sorted_clips = sorted(clips, key=lambda element: [element[0], element[1]])
        clips = sorted_clips
        n=len(clips)
        inf=102048900
        memo=[ [None for x in range(n)] for x in range(n) ] 

        @cache
        def fn(t, i):
            if i==-1:
                return 0 if t==0 else inf
                # if t==0:
                #     return 0
                # else:
                #     return inf
            ans=inf
            if clips[i][1]>=t:
                ans=min(ans, 1+fn(clips[i][0], i-1) )
            ans=min(ans, fn(t, i-1))
            return ans
        res = fn(time,n-1)
        return res if res!=inf else -1
    